Service overrides for step groups CDS-72217 v2 (#2743)
harness/developer-hub: docs/continuous-delivery/x-platform-cd-features/cd-steps/step-groups.md

`<+pipeline.stages.HTTP.spec.execution.steps.Group1.steps.OPTIONS.spec.url>`

## Override service variables in step groups

You can customize and override service variables during the execution of a step group. This provides significant flexibility and control over your pipelines. The expression `<+serviceVariableOverrides.VARIABLE_NAME>`overrides the original value of `VARIABLE_NAME`. You can use this expression in custom manifests, values YAML files, and other parts of the pipeline where variables are used.  

:::note important notes
- An overridden value is available only within the scope of the step group execution. In cases where a child group is included in a parent group, the child group override takes precedence. 

    The pipeline overrides variables based on the following priority:
    1. Step group override (child group, during execution)
    2. Step group override (parent group, during execution)
    3. Service environment overrides
    4. Environment configuration

>     5. Default value

Reviewer: Default value means the service variable value here. maybe we can put this as well

Reviewer 2: +1

> - Step group overrides are not recommended for artifacts and other objects fetched during the Service phase of a pipeline. The one exception is [custom remote manifests](/docs/continuous-delivery/deploy-srv-diff-platforms/kubernetes/cd-kubernetes-category/add-a-custom-remote-script-and-manifests/), because they can also be used during the execution stage of a pipeline.

Reviewer 2: We can mention manifests in general can be used with step group overrides with a mention of custom remote manifest.

<details><summary>Pipeline example: overriding variables in step groups</summary>

> The following pipeline illustrates how to override service variables. The service definition has two defined variables, `svar1` and `svar2`. The pipeline overrides these variables as follows:

Reviewer 2: A snapshot of pipeline on UI would help set the context clearer with what we are trying to do. I have added it in the jira. Please take a look.

`<+pipeline.stages.HTTP.spec.execution.steps.Group1.steps.OPTIONS.spec.url>`

## Override service variables in step groups

> You can customize and override service variables during the execution of a step group. This provides significant flexibility and control over your pipelines. The expression `<+serviceVariableOverrides.VARIABLE_NAME>`overrides the original value of `VARIABLE_NAME`. You can use this expression in custom manifests, values YAML files, and other parts of the pipeline where variables are used.  

Reviewer 3: ```suggestion
You can customize and override service variables during the execution of a step group. This provides significant flexibility and control over your pipelines. The expression `<+serviceVariableOverrides.VARIABLE_NAME>`overrides the original value of `VARIABLE_NAME`. You can use this expression in values YAML files and other parts of the pipeline where variables are used.  
```

Reviewer 3: not sure what a "custom manifest" is, but we only support expressions in values YAML files, not manifests.

> :::note important notes

Reviewer 3: ```suggestion
:::note Important notes
```

- An overridden value is available only within the scope of the step group execution. In cases where a child group is included in a parent group, the child group override takes precedence. 

    The pipeline overrides variables based on the following priority:
    1. Step group override (child group, during execution)
    2. Step group override (parent group, during execution)
    3. Service environment overrides
    4. Environment configuration
    5. Service variable value (default)

> - Step group overrides are not recommended for artifacts and other objects fetched during the Service phase of a pipeline. The one exception is [custom remote manifests](/docs/continuous-delivery/deploy-srv-diff-platforms/kubernetes/cd-kubernetes-category/add-a-custom-remote-script-and-manifests/), because they can also be used during the execution stage of a pipeline.

Reviewer 3: ```suggestion
- Step group overrides are not recommended for artifacts and other objects fetched in the **Service** section of a pipeline stage. The one exception is scripts in [custom remote manifests](/docs/continuous-delivery/deploy-srv-diff-platforms/kubernetes/cd-kubernetes-category/add-a-custom-remote-script-and-manifests/), because they can also be used during the **Execution** section of a pipeline stage.
```
